### PR TITLE
Support tuple type narrowing, fix call expression parser

### LINKED ISSUE #479: Object.freeze of a "as const" causes a Cannot read property 'map' of undefined
### Work as expected

When defined a typescript type from an array const, ts-json-schema-generator works as expected, and output the correct json schema.

`shared/src/foo.ts`
```ts
export const MEDIA_TYPES = ['image', 'video'] as const;

type MediaType = typeof MEDIA_TYPES[number];

export class Media {
	type?: MediaType
}
```

```sh
./node_modules/.bin/ts-json-schema-generator --path shared/src/foo.ts
```

Generate the correct json schema.

### Issue when using Object.freeze

However, if we try to freeze the array, Typescript works as expected, meaning the type gets inferred as the previous case, but ts-json-schema-generator, output a `Cannot read property 'map' of undefined` error

`shared/src/foo.ts`
```ts
export const MEDIA_TYPES = Object.freeze(['image', 'video'] as const);

type MediaType = typeof MEDIA_TYPES[number];

export class Media {
	type?: MediaType
}
```

```sh
./node_modules/.bin/ts-json-schema-generator --path shared/src/foo.ts

/....project..../node_modules/ts-json-schema-generator/dist/ts-json-schema-generator.js:45
        throw error;
        ^

TypeError: Cannot read property 'map' of undefined
    at CallExpressionParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/NodeParser/CallExpressionParser.js:19:67)
    at ChainNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:28:54)
    at TypeofNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeofNodeParser.js:31:45)
    at ChainNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:28:54)
    at IndexedAccessTypeNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/NodeParser/IndexedAccessTypeNodeParser.js:21:71)
    at ChainNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:28:54)
    at TypeAliasNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeAliasNodeParser.js:33:43)
    at AnnotatedNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/NodeParser/AnnotatedNodeParser.js:19:47)
    at ExposeNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/ExposeNodeParser.js:17:45)
    at CircularReferenceNodeParser.createType (/....project..../node_modules/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:21:43)
```

### Notes

In our code base, we need to export the possible values of those types for other reasons than just validation, but also need to make sure they are frozen to avoid accidental tampering. Typescript supports this transparently, and we were planning to use ts-json-schema-generator, but now we would need to add extra code to export the frozen array, and have the unfrozen one for internal typing.

### LINKED ISSUE #507: Incompatible with io-ts?
Hello there, I'm trying to use ts-json-schema-generator with io-ts, but currently it doesn't work well.

##### Work as expected

`string.ts`

```typescript
import * as t from "io-ts";
type MyType = t.TypeOf<typeof t.string>;

export class MyClass {
  public constructor(public type: MyType) {};
}
```

```bash
$ node_modules/.bin/ts-json-schema-generator -p string.ts
```

##### Broken

`string-broken.ts`

```typescript
import * as t from "io-ts";

const MyTypeT = t.string;
type MyType = t.TypeOf<typeof MyTypeT>;

export class MyClass {
  public constructor(public type: MyType) {};
}
```

```bash
$ node_modules/.bin/ts-json-schema-generator -p string-broken.ts
```

results to

```
Error: Unknown node "MyType" (ts.SyntaxKind = 198) at /Users/Prescott/Documents/Workspace/scratch/string-broken.ts(7,34)
```

### Object Type

`object.ts`

```typescript
import * as t from "io-ts";

const MyTypeT = t.type({ foo: t.string });
type MyType = t.TypeOf<typeof MyTypeT>;

export class MyClass {
  public constructor(public type: MyType) {};
}
```

```bash
$ node_modules/.bin/ts-json-schema-generator -p object.ts
```

results to

```
/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/ts-json-schema-generator.js:45
        throw error;
        ^

TypeError: Cannot read property 'map' of undefined
    at CallExpressionParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/NodeParser/CallExpressionParser.js:19:67)
    at ChainNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:28:54)
    at TypeofNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeofNodeParser.js:31:45)
    at ChainNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:28:54)
    at TypeReferenceNodeParser.createSubContext (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeReferenceNodeParser.js:51:51)
    at TypeReferenceNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeReferenceNodeParser.js:44:121)
    at ChainNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:28:54)
    at TypeAliasNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeAliasNodeParser.js:33:43)
    at AnnotatedNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/NodeParser/AnnotatedNodeParser.js:19:47)
    at ExposeNodeParser.createType (/Users/Prescott/Documents/Workspace/scratch/node_modules/ts-json-schema-generator/dist/src/ExposeNodeParser.js:17:45)
```

please let me know If you need further assistance :)

### PR DESCRIPTION
**Pull request by mooyoul**
##### Add PropertyAccessExpression support (w/ unit test)

Supports schema generation of below code:

```typescript
const FOO = { bar: "123"; }
export type MyType = typeof FOO.bar;
```

###### 
Support type Narrowing of tuple type using array-like object condition

Supports schema generation of below code:

```typescript
type Conditional<T extends any[]> = T extends { length: 2 } ? true : false;
export type TrueType = Conditional<[string, string]>; // true
export type FalseType = Conditional<[string, string, string]>; // false
export type FalseType = Conditional<string[]>; // false
```

##### Fix broken CallExpressionParser

Resolves issue with 3rd party type imports (e.g. io-ts)

## Related Issues

- Closes #507

### CODE REVIEW

**Comment by mooyoul** [comment]
Unfortunately, I have no more time to resolve these issues: 
- Compatibility issue with Zod (Related to #192)
- Support Inferred Type (Related to #513)

File: package.json [hunk 1]
```diff
@@ -51,6 +55,8 @@
     "eslint": "^7.8.1",
     "eslint-config-prettier": "^6.11.0",
     "eslint-plugin-prettier": "^3.1.4",
+    "fp-ts": "^2.8.2",
```
Comment (domoritz) [inline package.json:58]: Let's not rely on specific third-party libraries for the tests and instead boil the test case down to a unit test.

File: src/NodeParser/ArrayLiteralExpressionNodeParser.ts [hunk 2]
```diff
@@ -17,7 +16,7 @@ export class ArrayLiteralExpressionNodeParser implements SubNodeParser {
         if (node.elements) {
             const elements = node.elements.map((t) => this.childNodeParser.createType(t, context)).filter(notUndefined);
 
-            return new ArrayType(new UnionType(elements));
+            return new TupleType(elements);
```
Comment (domoritz) [inline src/NodeParser/ArrayLiteralExpressionNodeParser.ts:19]: Nice!

File: test/valid-data-third-party.test.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { assertValidSchema } from "./utils";
```
Comment (domoritz) [inline test/valid-data-third-party.test.ts:1]: Let's remove this test.

**Review by domoritz — APPROVED** [review]
Thank you! Let's remove the third party tests and boil down the problem to a unit test.

File: src/NodeParser/CallExpressionParser.ts [hunk 4]
```diff
@@ -14,10 +11,19 @@ export class CallExpressionParser implements SubNodeParser {
     }
     public createType(node: ts.CallExpression, context: Context): BaseType {
```
Comment (domoritz) [inline src/NodeParser/CallExpressionParser.ts:12]: Can you add a test that covers this code below?

File: src/NodeParser/CallExpressionParser.ts [hunk 5]
```diff
@@ -14,10 +11,19 @@ export class CallExpressionParser implements SubNodeParser {
...
+    }
+
+    private createSubContext(node: ts.CallExpression, parentContext: Context): Context {
+        const subContext = new Context(node);
 
-        // FIXME: make this general
-        return new TupleType([
-            new UnionType((type as any).typeArguments[0].types.map((t: any) => new LiteralType(t.value))),
-        ]);
+        node.arguments.forEach((arg) => {
```
Comment (domoritz) [inline src/NodeParser/CallExpressionParser.ts:23]: Please use `for (const arg of node.arguments) {}`

File: src/NodeParser/PropertyAccessExpressionParser.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import { BaseType } from "../Type/BaseType";
+
+export class PropertyAccessExpressionParser implements SubNodeParser {
+    public constructor(private typeChecker: ts.TypeChecker, private childNodeParser: NodeParser) {}
+
+    public supportsNode(node: ts.PropertyAccessExpression): boolean {
+        return node.kind === ts.SyntaxKind.PropertyAccessExpression;
+    }
+
+    public createType(node: ts.PropertyAccessExpression, context: Context): BaseType | undefined {
```
Comment (domoritz) [inline src/NodeParser/PropertyAccessExpressionParser.ts:13]: Please add a test that covers this case

**Review by domoritz — APPROVED** [review]
Awesome. Let's just fix a few more things and then I will merge the pull request and make a release.

**Comment by domoritz** [comment]
@mooyoul These changes are great and I would love to merge them. Can you finish up the two small changes?

**Comment by domoritz** [comment]
Hmm, unfortunately, this change breaks

```ts
const Foo = Object.keys({ foo: 42 });

export type MyType = typeof Foo[number];
```

```
Error: Unknown node "Object.keys({ foo: 42 })" (ts.SyntaxKind = 246)
```

**Comment by domoritz** [comment]
@mooyoul it would be awesome if you could help fix this case above.

**Pull request merged**